### DISCUSSION TITLE
CPM.cmake Install Fails Build with "-Og: not found"

### DISCUSSION THREAD

**Discussion by xanderdunn** [alandefreitas/matplotplusplus#164] [Q&A]
Ubuntu 18.04
GCC 10.3.0
cmake 3.20.1
CPM.cmake 0.32.2

I installed CPM.cmake with:
```bash
mkdir -p cmake
wget -O cmake/CPM.cmake https://github.com/cpm-cmake/CPM.cmake/releases/latest/download/get_cpm.cmake
```

I added matplotplusplus to my CMakeLists.txt with:
```cmake
include(cmake/CPM.cmake)
CPMAddPackage(
    NAME matplotplusplus
    GITHUB_REPOSITORY alandefreitas/matplotplusplus
    GIT_TAG origin/master # or whatever tag you want
)
option(CPM_USE_LOCAL_PACKAGES "Try `find_package` before downloading dependencies" ON)
```

Now, when I attempt to build my project I get:
```
[  1%] Building CXX object _deps/matplotplusplus-build/source/3rd_party/CMakeFiles/nodesoup.dir/nodesou
p/src/kamada_kawai.cpp.o
[  2%] Building CXX object _deps/matplotplusplus-build/source/3rd_party/CMakeFiles/nodesoup.dir/nodesou
p/src/nodesoup.cpp.o
[  4%] Building CXX object _deps/matplotplusplus-build/source/3rd_party/CMakeFiles/nodesoup.dir/nodesou
p/src/algebra.cpp.o
[  5%] Building CXX object _deps/matplotplusplus-build/source/3rd_party/CMakeFiles/nodesoup.dir/nodesou
p/src/fruchterman_reingold.cpp.o
[  7%] Building CXX object _deps/matplotplusplus-build/source/3rd_party/CMakeFiles/nodesoup.dir/nodesou
p/src/layout.cpp.o
g++-10: fatal error: no input files
compilation terminated.
g++-10: fatal error: no input files
compilation terminated.
/bin/sh: 1: -Og: not found
/bin/sh: 1: -Og: not found
g++-10: fatal error: no input files
compilation terminated.
_deps/matplotplusplus-build/source/3rd_party/CMakeFiles/nodesoup.dir/build.make:103: recipe for target
'_deps/matplotplusplus-build/source/3rd_party/CMakeFiles/nodesoup.dir/nodesoup/src/kamada_kawai.cpp.o'
failed
make[3]: *** [_deps/matplotplusplus-build/source/3rd_party/CMakeFiles/nodesoup.dir/nodesoup/src/kamada$
kawai.cpp.o] Error 127
make[3]: *** Waiting for unfinished jobs....
_deps/matplotplusplus-build/source/3rd_party/CMakeFiles/nodesoup.dir/build.make:131: recipe for target
'_deps/matplotplusplus-build/source/3rd_party/CMakeFiles/nodesoup.dir/nodesoup/src/nodesoup.cpp.o' fai$
ed
make[3]: *** [_deps/matplotplusplus-build/source/3rd_party/CMakeFiles/nodesoup.dir/nodesoup/src/nodeso$
p.cpp.o] Error 127
/bin/sh: 1: g++-10: fatal error: no input files
-Og: not found
compilation terminated.
_deps/matplotplusplus-build/source/3rd_party/CMakeFiles/nodesoup.dir/build.make:75: recipe for target $
_deps/matplotplusplus-build/source/3rd_party/CMakeFiles/nodesoup.dir/nodesoup/src/algebra.cpp.o' faile$
make[3]: *** [_deps/matplotplusplus-build/source/3rd_party/CMakeFiles/nodesoup.dir/nodesoup/src/algebr$
.cpp.o] Error 127
/bin/sh: 1: -Og: not found
g++-10: fatal error: no input files
compilation terminated.
_deps/matplotplusplus-build/source/3rd_party/CMakeFiles/nodesoup.dir/build.make:89: recipe for target $
_deps/matplotplusplus-build/source/3rd_party/CMakeFiles/nodesoup.dir/nodesoup/src/fruchterman_reingold$
cpp.o' failed
make[3]: *** [_deps/matplotplusplus-build/source/3rd_party/CMakeFiles/nodesoup.dir/nodesoup/src/frucht$
rman_reingold.cpp.o] Error 127
/bin/sh: 1: -Og: not found
_deps/matplotplusplus-build/source/3rd_party/CMakeFiles/nodesoup.dir/build.make:117: recipe for target
'_deps/matplotplusplus-build/source/3rd_party/CMakeFiles/nodesoup.dir/nodesoup/src/layout.cpp.o' faile$
make[3]: *** [_deps/matplotplusplus-build/source/3rd_party/CMakeFiles/nodesoup.dir/nodesoup/src/layout$
cpp.o] Error 127
CMakeFiles/Makefile2:341: recipe for target '_deps/matplotplusplus-build/source/3rd_party/CMakeFiles/n$
desoup.dir/all' failed
```
Where is the `-Og` flag coming from, and how might I fix this build? I see it listed [here](https://gcc.gnu.org/onlinedocs/gcc/Optimize-Options.html) as optimizing for the debug experience. Was it possibly removed in a more recent version of GCC?

**Comment by xanderdunn** [comment]
Note that I did not see this error when I copied the 1.1.0 source and built with this:
```
mkdir build
cd build
cmake ..
make -j 12
make install
```

Same GCC version: 10.3.0.

**Comment by alandefreitas** [comment]
Does 

```cmake
include(cmake/CPM.cmake)
CPMAddPackage(
    NAME matplotplusplus
    GITHUB_REPOSITORY alandefreitas/matplotplusplus
    GIT_TAG v1.1.0
)
option(CPM_USE_LOCAL_PACKAGES "Try `find_package` before downloading dependencies" ON)
```

work?

There might be something wrong with [this](https://github.com/alandefreitas/matplotplusplus/blob/abd69a4dfc316bebdca56f093371f75fee6e17be/cmake/functions/project_flags.cmake#L52) function.

**Reply by xanderdunn** [reply]
Thanks for the help. It looks like the change is to specify the v1.1.0 tag, correct? When I do that, I'm still seeing the same output:

<details>
[  4%] Building CXX object _deps/matplotplusplus-build/source/3rd_party/CMakeFiles/nodesoup.dir/nodesoup/src/kamada_kawai.cpp.o
[  5%] Building CXX object _deps/matplotplusplus-build/source/3rd_party/CMakeFiles/nodesoup.dir/nodesoup/src/fruchterman_reingold.cpp.o
[  7%] Building CXX object _deps/matplotplusplus-build/source/3rd_party/CMakeFiles/nodesoup.dir/nodesoup/src/nodesoup.cpp.o
g++-10: fatal error: no input files
compilation terminated.
/bin/sh: 1: -Og: not found
_deps/matplotplusplus-build/source/3rd_party/CMakeFiles/nodesoup.dir/build.make:117: recipe for target '_deps/matplotplusplus-build/source/3rd_party/CMakeFiles/nodesoup.dir/nodesoup/src/layout.cpp.o' failed
make[3]: *** [_deps/matplotplusplus-build/source/3rd_party/CMakeFiles/nodesoup.dir/nodesoup/src/layout.cpp.o] Error 127
make[3]: *** Waiting for unfinished jobs....
g++-10: fatal error: no input files
compilation terminated.
/bin/sh: 1: -Og: not found
_deps/matplotplusplus-build/source/3rd_party/CMakeFiles/nodesoup.dir/build.make:75: recipe for target '_deps/matplotplusplus-build/source/3rd_party/CMakeFiles/nodesoup.dir/nodesoup/src/algebra.cpp.o' failed
make[3]: *** [_deps/matplotplusplus-build/source/3rd_party/CMakeFiles/nodesoup.dir/nodesoup/src/algebra.cpp.o] Error 127
g++-10: fatal error: no input files
compilation terminated.
/bin/sh: 1: -Og: not found
_deps/matplotplusplus-build/source/3rd_party/CMakeFiles/nodesoup.dir/build.make:103: recipe for target '_deps/matplotplusplus-build/source/3rd_party/CMakeFiles/nodesoup.dir/nodesoup/src/kamada_kawai.cpp.o' failed
make[3]: *** [_deps/matplotplusplus-build/source/3rd_party/CMakeFiles/nodesoup.dir/nodesoup/src/kamada_kawai.cpp.o] Error 127
g++-10: fatal error: no input files
compilation terminated.
/bin/sh: 1: -Og: not found
_deps/matplotplusplus-build/source/3rd_party/CMakeFiles/nodesoup.dir/build.make:131: recipe for target '_deps/matplotplusplus-build/source/3rd_party/CMakeFiles/nodesoup.dir/nodesoup/src/nodesoup.cpp.o' failed
make[3]: *** [_deps/matplotplusplus-build/source/3rd_party/CMakeFiles/nodesoup.dir/nodesoup/src/nodesoup.cpp.o] Error 127
g++-10: fatal error: no input files
compilation terminated.
/bin/sh: 1: -Og: not found
_deps/matplotplusplus-build/source/3rd_party/CMakeFiles/nodesoup.dir/build.make:89: recipe for target '_deps/matplotplusplus-build/source/3rd_party/CMakeFiles/nodesoup.dir/nodesoup/src/fruchterman_reingold.cpp.o' failed
</details>

I did notice this at the beginning of my cmake configuration:
```
-- CPM: adding package matplotplusplus@1.1.0 (v1.1.0)
Setting CXX flags to default for Debug mode ( -D_GLIBCXX_USE_CXX11_ABI=1; -Og)
```
Which is from the `set_optimization_flags` macro you linked to, but I think the macro was introduced in v1.1.0

So I tried rolling back to `v1.0.1` rather than the latest `v1.1.0` . The build fails about half-way through with some unrelated compiler errors, but it no longer has the -Og problem:
<details>
/home/xander/dev/plutus_model/_deps/matplotplusplus-src/source/matplot/util/colors.h: In function ‘con$
texpr std::array<float, 4> matplot::default_color(size_t)’:
/home/xander/dev/plutus_model/_deps/matplotplusplus-src/source/matplot/util/colors.h:62:26: error: unu$
ed variable ‘num_default_colors’ [-Werror=unused-variable]
   62 |         constexpr size_t num_default_colors = 7;
      |                          ^~~~~~~~~~~~~~~~~~
In file included from /home/xander/dev/plutus_model/_deps/matplotplusplus-src/source/matplot/core/axes$
type.h:10,
                 from /home/xander/dev/plutus_model/_deps/matplotplusplus-src/source/matplot/core/axes_
object.cpp:6:
/home/xander/dev/plutus_model/_deps/matplotplusplus-src/source/matplot/util/colors.h: In function ‘cons
texpr std::array<float, 4> matplot::default_color(size_t)’:
/home/xander/dev/plutus_model/_deps/matplotplusplus-src/source/matplot/util/colors.h:62:26: error: unus
ed variable ‘num_default_colors’ [-Werror=unused-variable]
   62 |         constexpr size_t num_default_colors = 7;
      |                          ^~~~~~~~~~~~~~~~~~
</details>

**Comment by alandefreitas** [comment]
`-Og: not found` is a weird error.

Setting it to "O0" [here](https://github.com/alandefreitas/matplotplusplus/blob/abd69a4dfc316bebdca56f093371f75fee6e17be/cmake/functions/project_flags.cmake#L75) or to just nothing might solve the problem, but "-Og" should also work.
